### ISSUE TITLE
Request: Expose api for adding properties and passing flags between matchers and properties

### ISSUE THREAD

**Issue by stevezhu** [vitest-dev/vitest#2843]
### Clear and concise description of the problem

It's currently possible to extend `expect` in vitest by adding a plugin to chai directly, but it isn't possible to access the Assertion instance to pass flags between matchers and properties and it requires installing chai as a dependency separately.

This example shows how it's possible to add a custom chai property by doing something similar to `.resolves` and `.reject` currently, but is missing the features described above:
```typescript
import chai from 'chai';

chai.use((chai, utils) => {
  utils.addProperty(
    chai.Assertion.prototype,
    'custom',
    function __VITEST_CUSTOM__(this: Vi.Assertion) {
      const obj: unknown = utils.flag(this, 'object');

      // use obj

      // add flag
      utils.flag(this, 'custom', {
        a: 1,
      });

      return this;
    }
  );
});

expect.extend({
  toBeCustom(actual: unknown) {
    // no way to get the 'custom' flag here

    // matcher
    return {

    }
  },
});
```

This would be called like this.
```typescript
test('test', () => {
  expect(1).custom.toBeCustom();
})
```

### Suggested solution

Expose an api similar to `expect.extend` to add chai properties and also a way to pass flags maybe by passing the assertion instance into `MatcherState` or by passing a shared object per instance that is shared between matchers and properties that can be used to store flags.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by sheremet-va** [comment]
What is stopping you from using chai.use in Vitest?

**Comment by stevezhu** [comment]
> What is stopping you from using chai.use in Vitest?

As I mentioned in the issue, that is what I'm doing currently, but it's not possible to retrieve flags that are set between properties and matchers. And additionally syncing chai versions with vitest isn't the most optimal either.

Eg. A flag that is set at `.custom`, can't be accessed at `.toBeSomething()`.
```typescript
expect(1).custom.toBeSomething();
```

**Comment by sheremet-va** [comment]
> The current apis exposed by vitest don't seem sufficient to be able to implement something like expect(new Promise()).rejects.toThrow();, but definitely let me know if I'm missing something here.

Well, this API is not meant to support this. If you need assertion object, use chai.use, since it is a chai API. extend is Jest api, and the fact that they are on the dame object is just for better UX.

**Comment by stevezhu** [comment]
> TLDR: use one API, and don’t mix them

Yeah that's reasonable, I've switched to fully using the chai api.

Would be nice to eventually see this kind of api though to be able to extend vitest without installing chai.

**Comment by sheremet-va** [comment]
> > TLDR: use one API, and don’t mix them
> 
> 
> 
> Yeah that's reasonable, I've switched to fully using the chai api.
> 
> 
> 
> Would be nice to eventually see this kind of api though to be able to extend vitest without installing chai.
> 
> 

You don't need to install chai, Vitest exports its:

```ts
import { chai } from "vitest"
```

**Comment by stevezhu** [comment]
> You don't need to install chai, Vitest exports its:
> 
> ```ts
> import { chai } from "vitest"
> ```

Oh nice! I didn't know that vitest had that. Thanks! In that case my use case is fully covered.
